### ISSUE TITLE
text color on OS X

### ISSUE THREAD

**Issue by elvis-the-king** [XQuartz/XQuartz#294]
My app is built using the Fox-toolkit (fox-toolkit.org). It runs fine under Windows and Linux. However, on the Mac (OS X v12.6.1) the text in the FXTextBox control does not show up (only this control is messed up). It turns out the the text color is the same as the textbox background color so it does not show. Even if I hard code the text color it still does not show up. It seems to be always reset to the same color as the background color. I can see it when I highlight it with the mouse. I'm running XQuartz 2.8.2

**Comment by dpeterc** [comment]
I am also experiencing similar problem on my Motif applications, when user clicks on text field to edit it, current text becomes invisible and user does not see what are they writing.
Bug was introduced with XQuartz2.8.2beta2, beta1 or older versions were working OK.
Try installing this version of XQuartz to see if it helps:
https://www.xquartz.org/releases/XQuartz-2.8.2_beta1.html
If does, then we have stumbled in the same bug in XQuartz (or XFT or some other thing below).

It would be nice if this bug would be fixed, since Ventura brings new bugs, and we will not be able to run old XQuartz on new MacOS.

**Comment by elvis-the-king** [comment]
I just installed 2.8.3_rc1. It's the same. Sometimes when I restart my program it works fine but only about 10% of the time.

**Comment by dpeterc** [comment]
Sorry it did not work for you.
I will post tomorrow the pictures of how it looks in my app (correct behavior and wrong), so we can judge if it is really the same bug or not.
Is Fox toolkit using antialiased fonts (XFT)?

**Comment by elvis-the-king** [comment]
I'm not sure, it uses Xft and freetype libraries. Here are some screen shots. text_0.png is as it looks when it starts up. The textfields at the bottom should be filled. text_1.png shows when one is highlighted. text_2.png shows what happens sometimes, the text is visible but it is shifted to the right, it should be flush to the left size. When this happens you can;t see anything in narrow textboxes.
<img width="1518" alt="text_0" src="https://user-images.githubusercontent.com/98363967/202823176-f690bd63-421b-4acc-9ccb-0c93c6721e56.png">
<img width="1518" alt="text_1" src="https://user-images.githubusercontent.com/98363967/202823178-e28eacb6-c525-4fad-916b-1d10e067e4c1.png">
<img width="1518" alt="text_2" src="https://user-images.githubusercontent.com/98363967/202823183-50bf60d6-906a-4565-abbf-22d302d83e97.png">

**Comment by jeremyhu** [comment]
Could you narrow it down to whether it's a client bug or server bug and then bisect to figure out when this was introduced?

**Comment by elvis-the-king** [comment]
I have an old mac with version 1.20.10 on  it and it works fine on that. So something happened between 1.20.10 and 2.8.2. I know that a big span...

**Comment by elvis-the-king** [comment]
I can install some of the older versions and see when I get the problem. I earliest one I see is v2.1.0.1. Do I need to do them 1 by 1 or should I start with 1 in particular? They're a lot to do 1 by 1.

**Comment by jeremyhu** [comment]
1.20.10 is the version of xorg-server.
2.8.2 is the version of XQuartz.

XQuartz 2.8.1 had xorg-server-1.19.7 (plus patches)
XQuartz 2.8.2 had xorg-server-1.20.11 (plus patches)

**Comment by jeremyhu** [comment]
Can you please narrowing it down to client or server issue.  When you install an XQuartz package, it updates the client libraries (/opt/X11/lib) and the server (/Applications/Utilities/XQuartz.app).

You can backup XQuartz.app, eg:

```
cd /Applications/Utilities/
sudo cp -a XQuartz.app XQuartz-2.8.1.app
sudo tar cjf XQuartz-2.8.1.app.tar.bz2 XQuartz-2.8.1.app
rm -rf XQuartz-2.8.1.app
```

Then install the 2.8.2 update.
Then restore the backed up XQuartz-2.8.1.app:

```
cd /Applications/Utilities/
sudo tar xjf XQuartz-2.8.1.app.tar.bz2 XQuartz-2.8.1.app
```

Then launch XQuartz-2.8.1.app and see if the issue reproduces or not.  If it doesn't, then it means it was a server bug introduced in 1.20.x

**Comment by jeremyhu** [comment]
You can also try the 2.8.3 release candidates to see if the issue is resolved...

**Comment by elvis-the-king** [comment]
I just installed 2.8.3_rc2. It's the same, still work working correctly.

**Comment by elvis-the-king** [comment]
I followed the steps you outlined and the problem is still there. I guess its a client bug then.

**Comment by jeremyhu** [comment]
Ok, and to confirm, it's not an issue at all if you re-install 2.8.1, right?

Given that, can you nail down which client change (likely something in /opt/X11/lib) introduces the problem?  You can backup your newer /opt/X11 and then drop in updated dylibs until you observe the failure.

**Comment by elvis-the-king** [comment]
it's working now. A few days ago I installed 2.8.3_rc2 (server 21.1.4). I still had the problem. Today I just installed 2.8.1 without uninstalling 2.8.3_rc2 and its working. When I go to the "About X11" dialog It says I have 2.8.3_rc2 and 21.1.4. Then I reinstalled 2.8.3_rc2 (without uninstalling) and the problem is back. I then installed 2.8.2 and I still had the problem. Then I installed 2.8.1 and it is working fine. The "About X11" dialog still says 2.8.3_rc2. So something happened right after 2.8.1

**Comment by dpeterc** [comment]
I have researched this bug properly in my Motif application, so I will post a separate bug report, to avoid noise.
We actually have two bugs here - one is XQuartz reporting wrong version on downgrade, and other is a bug in Xft library, which is shipped with XOrg server, which is part of XQuartz.
Your Foxit based application is using XFT fonts, since the fonts on your screenshot are antialiased; Foxit web site mentions Xft font support, and doing antialiased fonts on X11 in other ways is quite difficult. I did not test the bug on Foxit application by myself.
Your temporary solution is downgrade of XQuartz to 2.8.2_beta3 or earlier, as you have already done, and final solution will be new release of XQuartz which will include updated version of Xft library.

**Comment by jeremyhu** [comment]
2.8.3_rc2 already has the latest libXft:

    xorg/lib/libXft libXft-2.3.4-56-g4356b58 (4356b583f8c23575a9ac25d49e7d224930107ef5)

Are you saying this is a known issue with the latest libXft?  Is there an upstream ticket for the issue (or a patch)?

**Comment by jeremyhu** [comment]
Actually, I updated all the libs in my wip branch for 2.8.4 prep.

Do you know which change in libXft fixes this?  Is there a way that I can reproduce this myself to test?

**Comment by jeremyhu** [comment]
Following this in #301.

**Comment by jeremyhu** [comment]
@elvis-the-king Could you please confirm this is fixed in [2.8.3_rc4](https://www.xquartz.org/releases/XQuartz-2.8.3_rc4.html)?

Thanks.

**Comment by elvis-the-king** [comment]
yes, it is working in 2.8.3_rc4. Thanks!
